tokio-epoll-uring/src/system/tests.rs

```diff
@@ -305,3 +311,144 @@ async fn test_write() {
+                    unreachable!()
+                }
+                _ = cancel.cancelled() => { drop(fut); fd }
+            }
+        });
+    }
+
+    for rx in submitted_or_enqueued {
+        rx.await.expect("test bug");
+    }
+
+    // all the futures have been submitted, drop them
+    cancel.cancel();
+    let mut timerfds = Vec::new();
+    while let Some(res) = tasks.join_next().await {
+        let timerfd = res.unwrap();
+        timerfds.push(timerfd);
+    }
+
+    // the slots are still blocked on the timerfd
+    // TODO: assert that directly
+    // assert it by starting a new read, it will enqueue
+    // TODO: use start_paused=true for this test, requires tokio upgrade
+    let fire_in = Duration::from_secs(1);
+    let fd = timerfd::oneshot(fire_in);
+    tokio::time::sleep(2 * fire_in).await;
+    let fut = timerfd::read(fd, system.clone());
+    let mut fut = std::pin::pin!(fut);
+    tokio::select! {
+        biased; // ensure future gets queued first
+        _ = &mut fut => {
+            panic!("future shouldn't be ready because all slots are still used")
+        }
+        _ = futures::future::ready(()) => { }
+    }
+
+    // unblock the tasks by firing their timerfds sooner, otherwise shutdown hangs forever
+    for timerfd in timerfds {
+        timerfd.set(Duration::from_millis(1));
+    }
+
+    // our read should complete because unblocking of the tasks
+    // frees up slots
+    let _: () = fut.await;
+
+    Arc::into_inner(system).unwrap().initiate_shutdown().await;
+}
+
+/// Scenario: a single tasks creates many futures that get submitted
+/// and hence occupy a slot, but the future never gets polled to completion,
+/// even though the io_uring-level operation has long completed.
+///
+/// The current behavior is that the operation waits for a slot to
+/// become available, i.e., it never completes.
+#[tokio::test]
+async fn test_slot_exhaustion_behavior_when_op_completes_but_future_does_not_get_polled() {
+    let system = Arc::new(System::launch().await.unwrap());
+
+    // Use up all slots.
+    let mut reads = FuturesUnordered::new();
+    let mut timerfds = Vec::new();
+    for _ in 0..RING_SIZE {
+        let oneshot = timerfd::oneshot(FOREVER);
+        let oneshot = Arc::new(oneshot);
+        let mut fut = Box::pin(tokio::task::unconstrained(timerfd::read(
+            oneshot.clone(),
+            system.clone(),
+        )));
+        let res = futures::poll!(&mut fut);
+        assert!(res.is_pending());
+        reads.push(fut);
+        timerfds.push(oneshot);
+    }
+
+    // An additional op will now wait forever for a free slot.
+    let mut nop = Box::pin(unconstrained(system.nop()));
+    tokio::select! {
+        biased; // ensure future gets queued first
+        res = &mut nop => {
+            panic!("nop shouldn't be able to get a slot because all slots are already used: {res:?}")
+        }
+        // TODO: use start_paused=true to de-flake this test
```

Comment:
If it's flaky then let's not merge it yet?

Reply:
Maybe flaky is not the right word.

It's not flaky.

But, take the scenario where we change the tokio-epoll-uring impl to allow for unbounded slots, with a delay for 2 seconds to avoid exhausting resources too quickly.
Then this test would be flaky because sometimes the `tokio::time::sleep` below this line would be ready before the nop, sometimes not.

The solution to this in tokio land is to use `#[tokio::test(start_paused=true)]` and in the line below not sleep for 2 seconds but FOREVER.

Then again, we have the background poller task waking up periodically by timeout to debug-log state, so, possibly `start_paused` doesn't work right away.

I'll just remove these comments.

Reply:
Actually I'll put a more detailed comment atop both test cases.